src/ripple/basics/LocalValue.h

```diff
@@ -0,0 +1,108 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_LOCALVALUE_H_INCLUDED
+#define RIPPLE_BASICS_LOCALVALUE_H_INCLUDED
+
+#include <boost/optional.hpp>
+#include <boost/thread/tss.hpp>
+#include <memory>
+#include <unordered_map>
+
+namespace ripple {
+
+namespace detail {
+
+struct LocalValues
+{
+    bool onCoro = true;
+
+    struct BasicValue
+    {
+        virtual ~BasicValue() = default;
+        virtual void* get() = 0;
+    };
+
+    template <class T>
+    struct Value : BasicValue
+    {
+        boost::optional<T> t;
+
+        void* get() override
+        {
+            return &t;
+        }
+    };
+
+    std::unordered_map<void const*, std::unique_ptr<BasicValue>> values;
+
+    static
+    inline
+    void
+    cleanup(LocalValues* lvs)
+    {
+        if (lvs && ! lvs->onCoro)
+            delete lvs;
+    }
+};
+
+template<class = void>
```

Comment:
I could be convinced that the definition for this function belongs in a .cpp file instead.
Rationale: The header `<boost/thread/tss.hpp>` is only needed for the definition of `getLocalValues`. Putting it in the .cpp file would save countless translation units from seeing unnecessary declarations.

I think we did it header only because most of the code has to be in the headers anyway, since `LocalValue` is a class template. So if we move `getLocalValues` we will end up with a .cpp file that has just one two line function in it. I think thats okay (preferred actually, now that I think about it).